### PR TITLE
Ryanunderhill/cuda shared

### PR DESCRIPTION
**Pull request by RyanUnderhill**
Please talk to me first before reviewing this change, as it requires background context. I'm mostly starting it right now to verify all of the build tests pass.

**Description**: Make the cuda provider be able to be dynamically loaded later at runtime - make it a shared library

**Motivation and Context**
With this change (similar to other shared providers) there can be one single onnxruntime library and it can be capable of loading multiple providers at runtime. Before this change, a special cuda onnxruntime would have to be built, after, it's the same onnxruntime, and cuda support is through the onnxruntime_providers_cuda.* shared library

### CODE REVIEW

File: java/src/main/java/ai/onnxruntime/OnnxRuntime.java [hunk 1]
```diff
@@ -111,12 +121,12 @@ static synchronized void init() throws IOException {
    *
    * @param file The file to remove.
    */
-  private static void cleanUp(File file) {
+  private static void cleanUp(File file, boolean on_exit_only) {
```
Comment (Craigacp) [inline java/src/main/java/ai/onnxruntime/OnnxRuntime.java:124]: Java variable names are idiomatically camelCase not snake_case. For both this and the load method.

File: java/src/main/java/ai/onnxruntime/OnnxRuntime.java [hunk 2]
```diff
@@ -201,11 +212,11 @@ private static void load(Path tempDirectory, String library) throws IOException
             os.write(buffer, 0, readBytes);
           }
         }
-        System.load(tempFile.getAbsolutePath());
+        if (system_load) System.load(tempFile.getAbsolutePath());
```
Comment (Craigacp) [inline java/src/main/java/ai/onnxruntime/OnnxRuntime.java:215]: I think this is the only inline if statement in the Java codebase. It should probably also include the logger message, and have an else branch that has a logging message at fine saying "library extracted" or something similar.

File: java/src/main/java/ai/onnxruntime/OnnxRuntime.java (hunk 1 above)
Comment (RyanUnderhill) [reply java/src/main/java/ai/onnxruntime/OnnxRuntime.java]: Sorry, I didn't think about it. Fixed

File: java/src/main/java/ai/onnxruntime/OnnxRuntime.java (hunk 2 above)
Comment (Craigacp) [reply java/src/main/java/ai/onnxruntime/OnnxRuntime.java]: Yeah, spotless does things like that. Our group in Oracle Labs always keeps braces around if statements, and so the ORT Java API is like that, and I didn't want it to stick out.

**Comment by Craigacp** [comment]
I'm not sure I completely understand the shared library loading process, but I think we probably want to refactor the Java bits a little. As I understand it, `onnxruntime_providers_shared` is necessary for all of the shared library providers, so we need to load it all the time (assuming it is present in the jar) and so it can be loaded with the core library, but `onnxruntime_providers_cuda` we could dependently extract just before the CUDA provider is added to the SessionOptions. That way we could throw an appropriate exception if we failed to extract the provider out of the jar file. Similarly we could do it for DNNL, OpenVINO and TensorRT. Having it extract a provider if that provider isn't going to be used seems like a waste of time.

One area I'm not sure I fully get is how it interacts with the `GetAvailableProviders` call in the C API. Does that trigger library loading? If not, how does it know what providers are available if they aren't compiled into the library? If `GetAvailableProviders` does trigger library loading then we'll have to catch that too and trigger shared library extraction.

**Comment by RyanUnderhill** [comment]
@Craigacp I agree it needs to be improved. It's certainly more optimal to dynamically extract the provider files needed vs doing them every time, but I wanted to do the simplest fix that made it work. Plus.. I'm not familiar with Java (as you can tell) and this should probably be done by someone more experienced in it. Once this change is in, it should be easy to fix.

**Comment by Craigacp** [comment]
> @Craigacp I agree it needs to be improved. It's certainly more optimal to dynamically extract the provider files needed vs doing them every time, but I wanted to do the simplest fix that made it work. Plus.. I'm not familiar with Java (as you can tell) and this should probably be done by someone more experienced in it. Once this change is in, it should be easy to fix.

Sure, I'm happy to do that, but I'm pretty low on time for the next few weeks while I work on a different release. I probably won't have time to do it before 1.8 is branched, but I've got a list of Java API fixes/features to add and I can put this on the list.

Is this going to land before 1.8?

File: cmake/CMakeLists.txt [hunk 3]
```diff
@@ -193,7 +193,8 @@ if(onnxruntime_USE_VALGRIND AND NOT WIN32)
 endif()
 
 if (onnxruntime_ENABLE_NVTX_PROFILE)
-  add_definitions(-DENABLE_NVTX_PROFILE=1)
+# TODO: This doesn't work with the shared cuda provider. Disabling temporarily to do a clean fix later as it wasn't trivial
+#  add_definitions(-DENABLE_NVTX_PROFILE=1)
```
Comment (edgchen1) [inline cmake/CMakeLists.txt:197]: should we warn or error out early (temporarily) instead if onnxruntime_ENABLE_NVTX_PROFILE is set?
otherwise it just silently doesn't work as advertised

Comment (RyanUnderhill) [reply cmake/CMakeLists.txt:197]: Warning would be good, if it errors it'll just break the pipeline that has it set.

File: include/onnxruntime/core/framework/tensor.h [hunk 4]
```diff
@@ -58,6 +58,9 @@ using BufferNakedPtr = void*;
 */
 class Tensor final {
  public:
+  static std::unique_ptr<Tensor> Create(MLDataType p_type, const TensorShape& shape, std::shared_ptr<IAllocator> allocator) { return std::make_unique<Tensor>(p_type, shape, allocator); }
```
Comment (edgchen1) [inline include/onnxruntime/core/framework/tensor.h:61]: nit: 120 character line length

File: java/src/main/java/ai/onnxruntime/OnnxRuntime.java [hunk 5]
```diff
@@ -111,12 +121,12 @@ static synchronized void init() throws IOException {
    *
    * @param file The file to remove.
    */
-  private static void cleanUp(File file) {
+  private static void cleanUp(File file, boolean onExitOnly) {
```
Comment (edgchen1) [inline java/src/main/java/ai/onnxruntime/OnnxRuntime.java:124]: >onExitOnly [](http://example.com/codeflow?start=49&length=10)

document new parameters
cleanUp() - onExitOnly, load() - systemLoad

File: onnxruntime/core/providers/shared_library/provider_bridge_provider.cc [hunk 6]
```diff
@@ -239,3 +312,172 @@ std::unique_ptr<OpKernelInfo> CopyOpKernelInfo(const OpKernelInfo& info) {
 }
 
 }  // namespace onnxruntime
+
+#include "core/providers/cpu/tensor/unsqueeze.h"
```
Comment (pranavsharma) [inline onnxruntime/core/providers/shared_library/provider_bridge_provider.cc:316]: What's the reason to include the operator headers here?

File: onnxruntime/contrib_ops/cuda/bert/attention.cc [hunk 7]
```diff
@@ -1,10 +1,9 @@
 // Copyright (c) Microsoft Corporation. All rights reserved.
 // Licensed under the MIT License.
 
-#include "attention.h"
-#include "core/framework/tensorprotoutils.h"
 #include "core/providers/cuda/cuda_common.h"
 #include "core/providers/cuda/shared_inc/fpgeneric.h"
+#include "attention.h"
```
Comment (edgchen1) [inline onnxruntime/contrib_ops/cuda/bert/attention.cc:6]: can the related header include be first?
[https://google.github.io/styleguide/cppguide.html#Names_and_Order_of_Includes](https://google.github.io/styleguide/cppguide.html#Names_and_Order_of_Includes)

File: onnxruntime/core/framework/provider_bridge_ort.cc [hunk 8]
```diff
@@ -514,45 +633,259 @@ struct ProviderHostImpl : ProviderHost {
...
+  int32_t* Tensor__MutableData_int32(Tensor* p) override { return p->MutableData<int32_t>(); }
+  uint32_t* Tensor__MutableData_uint32(Tensor* p) override { return p->MutableData<uint32_t>(); }
+  int64_t* Tensor__MutableData_int64(Tensor* p) override { return p->MutableData<int64_t>(); }
+  uint64_t* Tensor__MutableData_uint64(Tensor* p) override { return p->MutableData<uint64_t>(); }
   float* Tensor__MutableData_float(Tensor* p) override { return p->MutableData<float>(); }
+  double* Tensor__MutableData_double(Tensor* p) override { return p->MutableData<double>(); }
+  BFloat16* Tensor__MutableData_BFloat16(Tensor* p) override { return p->MutableData<BFloat16>(); }
+  MLFloat16* Tensor__MutableData_MLFloat16(Tensor* p) override { return p->MutableData<MLFloat16>(); }
+
+  const bool* Tensor__Data_bool(const Tensor* p) override { return p->Data<bool>(); }
```
[the remark this answers is not in the document]
Comment (edgchen1) [reply onnxruntime/core/framework/provider_bridge_ort.cc:706]: before we settle on this convention, note that it says here that "the identifiers with a double underscore anywhere are reserved"
[https://en.cppreference.com/w/cpp/language/identifiers](https://en.cppreference.com/w/cpp/language/identifiers)

File: cmake/CMakeLists.txt (hunk 3 above)
Comment (RyanUnderhill) [reply cmake/CMakeLists.txt:197]: Warning in place

File: include/onnxruntime/core/framework/tensor.h (hunk 4 above)
Comment (RyanUnderhill) [reply include/onnxruntime/core/framework/tensor.h:61]: Fixed

File: java/src/main/java/ai/onnxruntime/OnnxRuntime.java (hunk 5 above)
Comment (RyanUnderhill) [reply java/src/main/java/ai/onnxruntime/OnnxRuntime.java:124]: Fixed

File: onnxruntime/core/framework/provider_bridge_ort.cc (hunk 8 above)
[the remark this answers is not in the document]
Comment (RyanUnderhill) [reply onnxruntime/core/framework/provider_bridge_ort.cc:706]: Interesting, I thought it was just ones that start with double underscores. For here, I'm using __ in place of :: as a single underscore is less clear.

File: include/onnxruntime/core/providers/cuda/cuda_provider_factory.h [hunk 9]
```diff
@@ -4,6 +4,50 @@
 #include "onnxruntime_c_api.h"
```
[the remark this answers is not in the document]
Comment (RyanUnderhill) [reply include/onnxruntime/core/providers/cuda/cuda_provider_factory.h:4]: Yeah, this file is how our internal code interfaces with the Cuda provider (the ProviderInfo_CUDA interface is what the provider provides to us), but regular users should just use the C API instead of including this file for the OrtSessionOptionsAppendExecutionProvider_CUDA (ideally I'd just remove that function).

File: onnxruntime/core/framework/provider_bridge_ort.cc [hunk 10]
```diff
@@ -681,12 +1041,44 @@ std::shared_ptr<IExecutionProviderFactory> CreateExecutionProviderFactory_OpenVI
...
-const ProviderInfo_OpenVINO* GetProviderInfo_OpenVINO() {
+ProviderInfo_OpenVINO* GetProviderInfo_OpenVINO() {
   if (auto provider = s_library_openvino.Get())
-    return reinterpret_cast<const ProviderInfo_OpenVINO*>(provider->GetInfo());
+    return reinterpret_cast<ProviderInfo_OpenVINO*>(provider->GetInfo());
   return nullptr;
 }
 
+ProviderInfo_CUDA* GetProviderInfo_CUDA() {
+  if (auto provider = s_library_cuda.Get())
```
[the remark this answers is not in the document]
Comment (RyanUnderhill) [reply onnxruntime/core/framework/provider_bridge_ort.cc:1051]: Fixed them all to be consistent. I don't think it affects the code generated, it's just for clarity?

File: onnxruntime/core/providers/shared_library/provider_bridge_provider.cc (hunk 6 above)
Comment (RyanUnderhill) [reply onnxruntime/core/providers/shared_library/provider_bridge_provider.cc:316]: No reason now, I forget if there originally was one. I moved them to the top and made them more alphabetical

File: onnxruntime/contrib_ops/cuda/bert/attention.cc (hunk 7 above)
Comment (RyanUnderhill) [reply onnxruntime/contrib_ops/cuda/bert/attention.cc:6]: That one could happen to be included first. For others they needed to have the provider.api file first to define SHARED_PROVIDER.

Fixed

File: onnxruntime/core/providers/cuda/tensor/eye_like.cc [hunk 11]
```diff
@@ -17,20 +16,7 @@ ONNX_OPERATOR_KERNEL_EX(
...
-                                          DataTypeImpl::GetTensorType<int64_t>(),
-                                          DataTypeImpl::GetTensorType<int32_t>()})
-        .TypeConstraint("T2",
-                        std::vector<MLDataType>{
-                            DataTypeImpl::GetTensorType<float>(),
-                            DataTypeImpl::GetTensorType<double>(),
-                            DataTypeImpl::GetTensorType<uint64_t>(),
-                            DataTypeImpl::GetTensorType<int64_t>(),
-                            DataTypeImpl::GetTensorType<int32_t>()}),
+    (*KernelDefBuilder::Create()).TypeConstraint("T1", std::vector<MLDataType>{DataTypeImpl::GetTensorType<float>(), DataTypeImpl::GetTensorType<double>(), DataTypeImpl::GetTensorType<uint64_t>(), DataTypeImpl::GetTensorType<int64_t>(), DataTypeImpl::GetTensorType<int32_t>()}).TypeConstraint("T2", std::vector<MLDataType>{DataTypeImpl::GetTensorType<float>(), DataTypeImpl::GetTensorType<double>(), DataTypeImpl::GetTensorType<uint64_t>(), DataTypeImpl::GetTensorType<int64_t>(), DataTypeImpl::GetTensorType<int32_t>()}),
```
Comment (edgchen1) [inline onnxruntime/core/providers/cuda/tensor/eye_like.cc:19]: nit: line length

File: onnxruntime/test/framework/inference_session_test.cc [hunk 12]
```diff
@@ -1,5 +1,7 @@
 // Copyright (c) Microsoft Corporation. All rights reserved.
 // Licensed under the MIT License.
+#undef USE_CUDA  // TODO: Cuda is a shared library, so can't call any Cuda provider methods directly from here
```
Comment (edgchen1) [inline onnxruntime/test/framework/inference_session_test.cc:3]: >#undef USE_CUDA [](http://example.com/codeflow?start=0&length=15)

here and elsewhere, some tests are being disabled. is this absolutely necessary?

File: cmake/onnxruntime_unittests.cmake [hunk 13]
```diff
@@ -396,18 +396,6 @@ endif()
 
 set (onnxruntime_test_providers_dependencies ${onnxruntime_EXTERNAL_DEPENDENCIES})
 
-if(onnxruntime_USE_CUDA)
-  list(APPEND onnxruntime_test_providers_dependencies onnxruntime_providers_cuda)
-endif()
-
-if(onnxruntime_USE_DNNL)
```
Comment (jywu-msft) [inline cmake/onnxruntime_unittests.cmake:403]: why these changes for DNNL and OPENVINO ?

File: cmake/onnxruntime_python.cmake [hunk 14]
```diff
@@ -370,6 +369,7 @@ add_custom_command(
 )
 
 if (NOT onnxruntime_MINIMAL_BUILD AND NOT onnxruntime_EXTENDED_MINIMAL_BUILD 
+                                  AND NOT onnxruntime_ENABLE_TRAINING
```
Comment (jywu-msft) [inline cmake/onnxruntime_python.cmake:372]: why is this if block needed?
seems like onnxruntime_providers_shared is copied for the various EP cases line 430-468 below?

File: include/onnxruntime/core/providers/cuda/cuda_provider_factory.h (hunk 9 above)
[the remark this answers is not in the document]
Comment (pranavsharma) [reply include/onnxruntime/core/providers/cuda/cuda_provider_factory.h:4]: Would it be better to move the interface functions to a separate file called cuda_provider_interfaces? Besides we can't remove OrtSessionOptionsAppendExecutionProvider_CUDA since the CUDA DLL is published and removing it would break ABI.

File: onnxruntime/core/framework/provider_bridge_ort.cc (hunk 8 above)
[the remark this answers is not in the document]
Comment (pranavsharma) [reply onnxruntime/core/framework/provider_bridge_ort.cc:706]: Can you document somewhere what these new conventions mean?

File: onnxruntime/core/framework/provider_bridge_ort.cc (hunk 10 above)
[the remark this answers is not in the document]
Comment (pranavsharma) [reply onnxruntime/core/framework/provider_bridge_ort.cc]: Yes, for clarity. Also, adherence to coding conventions.

File: onnxruntime/test/framework/cuda/fence_cuda_test.cc [hunk 15]
```diff
@@ -1,5 +1,6 @@
 // Copyright (c) Microsoft Corporation. All rights reserved.
 // Licensed under the MIT License.
+#if 0  // TODO: Can't call these directly from external code as Cuda is now a shared library
```
Comment (pranavsharma) [inline onnxruntime/test/framework/cuda/fence_cuda_test.cc:3]: How do we re-enable these tests? I've not seen other files, but a bit concerned that we're disabling these tests.

File: onnxruntime/core/providers/cpu/controlflow/utils.h [hunk 16]
```diff
@@ -7,12 +7,14 @@
 #include <string>
 #include <vector>
 
-#include "core/common/common.h"
 #include "core/framework/feeds_fetches_manager.h"
+#ifdef SHARED_PROVIDER
```
Comment (pranavsharma) [inline onnxruntime/core/providers/cpu/controlflow/utils.h:11]: It's not clear why I would check for this macro in the implementation of a CPU op. Is this documented somewhere?

File: cmake/onnxruntime_unittests.cmake (hunk 13 above)
Comment (RyanUnderhill) [reply cmake/onnxruntime_unittests.cmake:403]: I noticed these extraneous dependencies were here, so removed them for the other shared providers too.

File: cmake/onnxruntime_python.cmake (hunk 14 above)
Comment (RyanUnderhill) [reply cmake/onnxruntime_python.cmake:372]: This block was added for Cheng's text_execution_provider (a custom external provider) to ensure that 'onnxruntime_providers_shared' existed in his case. With training enabled, his test_execution_provider fails to build the provider bridge (as the bridge does depend on training), so the simple fix was to just disable it when training was enabled. I talked to Cheng about it and he recommended this as the fix (plus disabling the python test when training is enabled).

File: onnxruntime/test/framework/cuda/fence_cuda_test.cc (hunk 15 above)
Comment (RyanUnderhill) [reply onnxruntime/test/framework/cuda/fence_cuda_test.cc:3]: I need to expose all of the internal methods they call through the ProviderInfo_CUDA() interface so that they can still access them. It's at least another couple weeks of work and I need to do it for other providers also.

The ProviderInfo_CUDA() style interface was created to solve this problem, it's just not yet converted over as it wasn't trivial to change.

File: onnxruntime/core/providers/cpu/controlflow/utils.h (hunk 16 above)
Comment (RyanUnderhill) [reply onnxruntime/core/providers/cpu/controlflow/utils.h:11]: Many of the cuda ops will share functions with the cpu ops, so they '#include' the cpu headers to access these methods, which will fail on the internal headers that don't work from a shared provider.

Typically it's either a base class, a structure, or parameter validation methods the cuda providers share with the cpu one. These functions are routed across the bridge.

File: onnxruntime/test/framework/inference_session_test.cc (hunk 12 above)
Comment (RyanUnderhill) [reply onnxruntime/test/framework/inference_session_test.cc:3]: (we discussed this outside of github, but answering here for history) Yep, I'm logging a work item to rewrite the tests to work across the shared library boundary, the existing ones all call directly into the code and it'll take a few more weeks to translate them.

File: onnxruntime/core/providers/cuda/tensor/eye_like.cc (hunk 11 above)
Comment (RyanUnderhill) [reply onnxruntime/core/providers/cuda/tensor/eye_like.cc:19]: Fixed, I wonder if clang-format did that for some reason.

File: onnxruntime/core/providers/cpu/controlflow/utils.h (hunk 16 above)
Comment (pranavsharma) [reply onnxruntime/core/providers/cpu/controlflow/utils.h:11]: Sure. Would be good to comment this somewhere so that people are not left guessing.

File: onnxruntime/test/framework/cuda/fence_cuda_test.cc (hunk 15 above)
Comment (pranavsharma) [reply onnxruntime/test/framework/cuda/fence_cuda_test.cc:3]: Is there an easy way to grep for the tests that have been disabled due to this change? This will make it easier to find these tests and quantify.

File: onnxruntime/core/providers/shared_library/provider_bridge_provider.cc (hunk 6 above)
Comment (pranavsharma) [reply onnxruntime/core/providers/shared_library/provider_bridge_provider.cc]: I meant why do we need to include the operator headers in this bridge code?

File: onnxruntime/core/framework/provider_bridge_ort.cc [hunk 17]
```diff
@@ -514,45 +633,259 @@ struct ProviderHostImpl : ProviderHost {
...
+  int32_t Tensor__GetElementType(const Tensor* p) override { return p->GetElementType(); }
+  MLDataType Tensor__DataType(const Tensor* p) override { return p->DataType(); }
+
+  // AllocatorManager (direct)
+  void AllocatorManager__InsertAllocator(AllocatorManager* p, AllocatorPtr allocator) override { p->AllocatorManager::InsertAllocator(allocator); }
+  AllocatorPtr AllocatorManager__GetAllocator(const AllocatorManager* p, int id, OrtMemType mem_type) override { return p->AllocatorManager::GetAllocator(id, mem_type); };
+
+#ifdef USE_CUDA
+  // GatherElements (direct)
+  Status GatherElements__ValidateInputShapes(const TensorShape& input_data_shape, const TensorShape& indices_shape, int64_t axis) override { return GatherElements::ValidateInputShapes(input_data_shape, indices_shape, axis); }
```
Comment (pranavsharma) [inline onnxruntime/core/framework/provider_bridge_ort.cc:756]: We might want to clean this up. Having operator code here seems very weird.

**Review by pranavsharma — APPROVED** [review]
Since this is a big PR and all pipelines are passing let's address the comments in a separate PR.

File: cmake/onnxruntime_java.cmake [hunk 18]
```diff
@@ -134,6 +134,10 @@ if (WIN32)
   if(NOT onnxruntime_ENABLE_STATIC_ANALYSIS)
     add_custom_command(TARGET onnxruntime4j_jni POST_BUILD COMMAND ${CMAKE_COMMAND} -E create_symlink $<TARGET_FILE:onnxruntime> ${JAVA_PACKAGE_LIB_DIR}/$<TARGET_FILE_NAME:onnxruntime>)
     add_custom_command(TARGET onnxruntime4j_jni POST_BUILD COMMAND ${CMAKE_COMMAND} -E create_symlink $<TARGET_FILE:onnxruntime4j_jni> ${JAVA_PACKAGE_JNI_DIR}/$<TARGET_FILE_NAME:onnxruntime4j_jni>)
+    if (onnxruntime_USE_CUDA)
```
Comment (Craigacp) [inline cmake/onnxruntime_java.cmake:137]: This block is only in the WIN32 branch, and needs to be in the else case as well, with the appropriate substitution of `TARGET_FILE_NAME` for `TARGET_LINKER_FILE_NAME`. Without it it doesn't bundle the cuda and shared provider libraries into the jar on Linux.

Comment (RyanUnderhill) [reply cmake/onnxruntime_java.cmake:137]: Ah, I see it. It's unfortunate that these have to be duplicated with a slight change for multiple platforms.

Comment (Craigacp) [reply cmake/onnxruntime_java.cmake:137]: I'm going to have to refactor the Java cmake bits a little anyway to add in support for the other shared library providers, so provided there is time to do it before this code is released we're ok.

Comment (Craigacp) [reply cmake/onnxruntime_java.cmake:137]: It might be that I don't understand how CMake works well enough to get around it, but when I looked into it last year this was the simplest way I could figure out.

Comment (RyanUnderhill) [reply cmake/onnxruntime_java.cmake:137]: No worries, I pushed a change just a minute ago with what you said and made sure to do the substitution. Cmake is certainly nebulous... and not easy to know when you did something wrong.

File: cmake/onnxruntime_python.cmake (hunk 14 above)
Comment (snnn) [reply cmake/onnxruntime_python.cmake:372]: So, when training is enable, how will the CUDA EP work if it can't find the bridge shared library?

**Comment by snnn** [comment]
>  * Copyright (c) 2019, 2020, Oracle and/or its affiliates. All rights reserved.

From the Linux GPU CI pipeline's build log I saw " InferenceTest > testCUDA() PASSED". That's great. Thank you!


---
Refers to: java/src/main/java/ai/onnxruntime/OnnxRuntime.java:2 in c569dee. [](commit_id = c569dee80a298292f00869ecd87c8a092c85b49d, deletion_comment = False)

File: onnxruntime/core/framework/provider_bridge_ort.cc [hunk 19]
```diff
@@ -593,14 +927,15 @@ struct ProviderSharedLibrary {
...
-bool InitProvidersSharedLibrary(){
+bool InitProvidersSharedLibrary() {
   return s_library_shared.Ensure();
 }
 
 struct ProviderLibrary {
-  ProviderLibrary(const char* filename) : filename_{filename} {}
-  ~ProviderLibrary() { /*assert(!handle_);*/
-  }                    // We should already be unloaded at this point (disabled until Python shuts down deterministically)
+  ProviderLibrary(const char* filename, bool unload = true) : filename_{filename}, unload_{unload} {}
```
Comment (snnn) [inline onnxruntime/core/framework/provider_bridge_ort.cc:935]: Can you change "const char*" to "const ORTCHAR_T*" instead?

**Comment by snnn** [comment]
>     Env::Default().GetSymbolFromLibrary(handle_, "GetProvider", (void**)&PGetProvider);

Please check the returned value

---
Refers to: onnxruntime/core/framework/provider_bridge_ort.cc:955 in c569dee. [](commit_id = c569dee80a298292f00869ecd87c8a092c85b49d, deletion_comment = False)

File: onnxruntime/core/framework/provider_bridge_ort.cc [hunk 20]
```diff
@@ -593,14 +927,15 @@ struct ProviderSharedLibrary {
 
 static ProviderSharedLibrary s_library_shared;
```
Comment (snnn) [inline onnxruntime/core/framework/provider_bridge_ort.cc:928]: Is it possbile to let OrtEnv manage these static varibles? I worry the destruct order of s_library_shared/s_library_xxx and something else.

Comment (RyanUnderhill) [reply onnxruntime/core/framework/provider_bridge_ort.cc:928]: Fear not, they might be static variables but they are destroyed by calling 'UnloadSharedProviders()' to guarantee the destruction order. Their destructors do nothing.

**Comment by Craigacp** [comment]
I looked at the libraries linked by libonnxruntime.so from this PR, and it's still linking against CUDA. I thought this change was supposed to remove the CUDA dependency from the main lib?

```
ldd libonnxruntime.so
	linux-vdso.so.1 =>  (0x00007fff191d9000)
	libcublas.so.11 => <path-to-cuda>/libcublas.so.11 (0x00007fb68ab69000)
	libcudnn.so.8 => <path-to-cuda>/libcudnn.so.8 (0x00007fb68a940000)
	libcurand.so.10 => <path-to-cuda>/libcurand.so.10 (0x00007fb685a10000)
	libcufft.so.10 => <path-to-cuda>/libcufft.so.10 (0x00007fb6776ac000)
	libdl.so.2 => /lib64/libdl.so.2 (0x00007fb6774a8000)
	libcupti.so.11.1 => <path-to-cuda>/libcupti.so.11.1 (0x00007fb676c2d000)
	librt.so.1 => /lib64/librt.so.1 (0x00007fb676a25000)
	libpthread.so.0 => /lib64/libpthread.so.0 (0x00007fb676809000)
	libstdc++.so.6 => /lib64/libstdc++.so.6 (0x00007fb676501000)
	libm.so.6 => /lib64/libm.so.6 (0x00007fb6761ff000)
	libgcc_s.so.1 => /lib64/libgcc_s.so.1 (0x00007fb675fe9000)
	libc.so.6 => /lib64/libc.so.6 (0x00007fb675c1b000)
	/lib64/ld-linux-x86-64.so.2 (0x00007fb693bc2000)
	libcublasLt.so.11 => <path-to-cuda>/libcublasLt.so.11 (0x00007fb667c27000)
	libutil.so.1 => /lib64/libutil.so.1 (0x00007fb667a24000)
```

I was hoping to close #7044 after this is merged in (by making the loading conditional on a user adding the CUDA provider to their session options), but I can't if libonnxruntime still links to CUDA.

**Comment by RyanUnderhill** [comment]
@Craigacp What build options did you use? I do see an unexpected dependency on cupti was added recently, but on my windows build, onnxruntime.dll doesn't load all of the cuda libraries until I load the providers_shared_cuda.dll file. There could be a build option that adds a cuda dependency that I didn't see.

**Comment by Craigacp** [comment]
@RyanUnderhill This was my build command: `./build.sh --config Release --build --build_java --parallel --cuda_home <path-to-cuda> --cudnn_home <path-to-cudnn> --update --cmake_path /usr/bin/cmake3 --use_cuda`.

**Comment by RyanUnderhill** [comment]
@Craigacp Ah, we think we figured it out. It looks like it was the cupti dependency. We rebuilt on linux with my latest changes and ldd shows no cuda dependencies.

**Comment by Craigacp** [comment]
Looks like the fix for cupti was backed out? So it won't be possible to run the GPU compiled binary on Linux machines without CUDA in 1.8?

**Comment by jywu-msft** [comment]
> Looks like the fix for cupti was backed out? So it won't be possible to run the GPU compiled binary on Linux machines without CUDA in 1.8?

No, we still aim to get this in for 1.8
CUPTI integration with ORT itself was backed out by https://github.com/microsoft/onnxruntime/pull/7763
so the CUPTI fix is not needed anymore.

**Review by jywu-msft — APPROVED** [review]

**Comment by Craigacp** [comment]
Even after this has been merged into master I'm still seeing the CUDA libraries being linked into `libonnxruntime.so`.

```
ldd build/Linux/Release/libonnxruntime.so.1.8.0 
	linux-vdso.so.1 =>  (0x00007ffc52d6a000)
	libcublas.so.11 => <path-to-cuda>/libcublas.so.11 (0x00007f98897ee000)
	libcudnn.so.8 => <path-to-cuda>/libcudnn.so.8 (0x00007f98895c5000)
	libcurand.so.10 => <path-to-cuda>/libcurand.so.10 (0x00007f9884695000)
	libcufft.so.10 => <path-to-cuda>/libcufft.so.10 (0x00007f9876331000)
	libdl.so.2 => /lib64/libdl.so.2 (0x00007f987612d000)
	librt.so.1 => /lib64/librt.so.1 (0x00007f9875f25000)
	libpthread.so.0 => /lib64/libpthread.so.0 (0x00007f9875d09000)
	libstdc++.so.6 => /lib64/libstdc++.so.6 (0x00007f9875a01000)
	libm.so.6 => /lib64/libm.so.6 (0x00007f98756ff000)
	libgcc_s.so.1 => /lib64/libgcc_s.so.1 (0x00007f98754e9000)
	libc.so.6 => /lib64/libc.so.6 (0x00007f987511b000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f9892845000)
	libcublasLt.so.11 => <path-to-cuda>/libcublasLt.so.11 (0x00007f9867127000)
```

I used the same build command as before, and rm'd the build folder before compiling everything.

**Comment by jywu-msft** [comment]
> @Craigacp Ah, we think we figured it out. It looks like it was the cupti dependency. We rebuilt on linux with my latest changes and ldd shows no cuda dependencies.

turns out that wasn't the case.  even after removing CUPTI dependency. CUDA dependency is still there. It seems to be happening on CentOS but not Ubuntu.
